sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp

```diff
@@ -0,0 +1,452 @@
+        for (int j = 0; j < 4; ++j) {
+          for (int i = 0; i < 4; ++i) {
+            const int d_idx =
+                thread_x + i * stride + thread_y * 4 * stride + j * 8 * stride;
+            dst[d_idx] = src.data[i + 4 * j];
+          }
+        }
+      } else {
+        for (int j = 0; j < 4; ++j) {
+          for (int i = 0; i < 4; ++i) {
+            const int d_idx = i + thread_x * stride + thread_y * 4 + j * 8;
+            dst[d_idx] = src.data[i + 4 * j];
+          }
+        }
+      }
+    }
+  }
+}
+
+template <typename Group, typename T, size_t M, size_t N,
+          access::address_space Space, access::decorated IsDecorated>
+void joint_matrix_store_hip(
+    const joint_matrix_hip<
+        T, sycl::ext::oneapi::experimental::matrix::use::accumulator, M, N,
+        sycl::ext::oneapi::experimental::matrix::layout::dynamic> &src,
+    multi_ptr<T, Space, IsDecorated> dst, size_t stride,
+    sycl::ext::oneapi::experimental::matrix::layout layout, Group &sg) {
+  if (sycl::ext::oneapi::experimental::matrix::layout::row_major == layout) {
+    store_layoutT<Group,
+                  sycl::ext::oneapi::experimental::matrix::layout::row_major>(
+        src, dst, stride, sg);
+  } else {
+    store_layoutT<Group,
+                  sycl::ext::oneapi::experimental::matrix::layout::col_major>(
+        src, dst, stride, sg);
+  }
+}
+
+template <typename Tm, typename Tc, std::size_t M, std::size_t K, std::size_t N,
+          sycl::ext::oneapi::experimental::matrix::layout LayoutA,
+          sycl::ext::oneapi::experimental::matrix::layout LayoutB>
+void joint_matrix_mad_hip(
+    joint_matrix_hip<
+        Tc, sycl::ext::oneapi::experimental::matrix::use::accumulator, M, N,
+        sycl::ext::oneapi::experimental::matrix::layout::dynamic> &D,
+    const joint_matrix_hip<Tm, sycl::ext::oneapi::experimental::matrix::use::a,
+                           M, K, LayoutA> &A,
+    const joint_matrix_hip<Tm, sycl::ext::oneapi::experimental::matrix::use::b,
+                           K, N, LayoutB> &B,
+    const joint_matrix_hip<
+        Tc, sycl::ext::oneapi::experimental::matrix::use::accumulator, M, N,
+        sycl::ext::oneapi::experimental::matrix::layout::dynamic> &C) {
+  if constexpr (std::is_same_v<Tm, sycl::half>) {
+    if constexpr (M == 16 && N == 16) {
+      auto result = __builtin_amdgcn_mfma_f32_16x16x16f16(
+          *reinterpret_cast<const float16x4 *>(&A.data),
+          *reinterpret_cast<const float16x4 *>(&B.data),
+          *reinterpret_cast<const floatx4 *>(&C.data), 0, 0, 0);
+      for (int i = 0; i < 4; ++i)
+        D.data[i] = result[i];
+    } else if constexpr (M == 32 && N == 32) {
+      auto result = __builtin_amdgcn_mfma_f32_32x32x8f16(
+          *reinterpret_cast<const float16x4 *>(&A.data),
+          *reinterpret_cast<const float16x4 *>(&B.data),
+          *reinterpret_cast<const floatx16 *>(&C.data), 0, 0, 0);
+      for (int i = 0; i < 16; ++i)
+        D.data[i] = result[i];
```

Comment:
Would it be more efficient to use `std::memcpy` here?

Reply:
I am not sure that would be better in terms of performance. `std::memcpy` takes the number of `bytes` to be copied and copies `count` bytes form src to dest. My impression is that iterating `N` timers over `Type`  should not be slower than iterating iterating `N*sizeof(Type)` times over `bytes`